### ISSUE TITLE
arcgis rasters throw PILLOW error

### ISSUE THREAD

**Issue by bobombolo** [bobombolo/preload#7]
<pre>
Traceback (most recent call last):
  File "./preload.py", line 105, in <module>
    elev = im_elev.getpixel((imx, imy))
  File "/usr/lib/python3/dist-packages/PIL/Image.py", line 1210, in getpixel
    self.load()
  File "/usr/lib/python3/dist-packages/PIL/ImageFile.py", line 248, in load
    if not self.map and (not LOAD_TRUNCATED_IMAGES or t == 0) and e < 0:
TypeError: unorderable types: tuple() < int()

</pre>

**Comment by bobombolo** [comment]
might want to switch to wand, which has the added bonus of surely not opening the whole file into RAM and also of being able to get an array of a set of pixels instead of individually getting pixels:

comment from Gael:

To get a pixel with wand, use `image[x,y]`. But that's slow. Better split the image with `image[xmin:xmax,ymin:ymax]`, get raw data using `make_blob("grey")`, and store it in an array with the `array` library: `array("h", blob)`.
See [this code](https://github.com/Gael-de-Sailly/rivermapper/blob/master/rivermapper.py#L17-L41).

**Comment by Tank-ACiD** [comment]
f-arcgis! lets stick to open source produced/saved rasters only? It
probably has to do with the compression used in ArcGIS versus the
non-compression used in qgis etc.

**Comment by bobombolo** [comment]
ya I was just testing. that was with Pillow, maybe Wand for py can do it, but anyway I am running the script on the mexico rasters (small, qgis) you provided, been running all night and will be done in an hour or so. Should give us an idea of worst case time...

**Comment by bobombolo** [comment]
as soon as I watch streaming video the speed plummets haha

**Comment by Tank-ACiD** [comment]
i'll take a look at the current master of preload today, see how it runs on
my machine and what the output is like.

**Comment by bobombolo** [comment]
make sure you update libminetest, which I am including in the preload repo...
you might want to give me a few hours to solve #6

**Comment by Tank-ACiD** [comment]
got and built the latest python libminetest from Mag,
got preload and followed the instructions,
line 46: map_interface.container.empty_map()
AttributeError: 'MapInterface' object has no attribute 'container'

**Comment by bobombolo** [comment]
sounds like you don't have the latest libminetest, the field 'container' used to be 'interface', not sure that the setup.py works or whatever... check your c:\python34\Lib\libminetest\map.py for the definition of the MapInterface and see if the field is called interface or container?
possibly delete the libminetest from the Lib folder and rely on the script to use the included lib which I will always keep synced with the script itself?

**Comment by Tank-ACiD** [comment]
Even copied the mod supplied libminetest files into my site packages for
python, and ran script, same, commented it out since it wasn't in the last
version, then ran into

line 130: pos = libminetest.utils.Pos(imx, elev, -imy)
TypeError: __init__() takes from 1 to 2 positional arguments but 4 were
given

**Comment by Tank-ACiD** [comment]
looks good to me...

class MapInterface:
def __init__(self, datafile, backend = "sqlite3"):
self.datafile = datafile
self.container = MapVessel(datafile, backend)
self.max_cache_size = 100
self.mapblocks = dict()
self.cache_history = StackCache()
self.mod_cache = StackCache()

**Comment by bobombolo** [comment]
does the python cache folders in the Lib folder and parent folder have anything to do with it? because that also sounds like a difference from the old version...

**Comment by bobombolo** [comment]
the old version took a single argument {x,y,z} and the new version takes three arguments. You are definately still using the old one... somehow. yay package management. Delete the cache folders and libminetest in c:\python34 ?

**Comment by Tank-ACiD** [comment]
no cache, probable an issue, i got the latest form Mag and did a build,
then manually copied the build dir, this time i ran install, it installed
and egg and then things are running smoothly. I should have gone through
and reduced the amount of output, every print slows the process way down :)
200,000 pixels in 154 sec

**Comment by Tank-ACiD** [comment]
oh and im also upgrading java at the same time...my bad...

**Comment by Tank-ACiD** [comment]
I'd also like to see the landscape artifact generation controlled by a flag
so if i want a barren landscape I  can...I know all that can be tweaked
down the road, but it will speed things up ever so slightly.

**Comment by bobombolo** [comment]
like you mean trees and whatnot? this will all be controlled as per #5 and I am planning to make a GUI or a user prompt for options at the CLI

**Comment by Tank-ACiD** [comment]
based on the first 100k i est that the small 8 bit well relatively small 6k
x 6k image that i supplied will take me a little over 8 hours to generate
with this script

**Comment by Tank-ACiD** [comment]
just for ref again on this thread:
Intel 8 core @ 2.66GHz
Win 7
Ram-Drive FB-DIMM PC2-5300 @ 667mhz

1M pixels @ 940 sec

so est. 9.88 hrs to complete

**Comment by Tank-ACiD** [comment]
ok, about to head to work for a few hours will pick this back up after :)

**Comment by bobombolo** [comment]
i have made the trees flag and toned down the reporting, will post soon...

**Comment by bobombolo** [comment]
i just finished (with last night's code) preloading your rasters: 

<pre>
[BLOCK: (384, 131, -385)] [39627.81s] [READ: 37871716 pixels] [WROTE: 37761081 voxels]
</pre>

![screenshot_20160407_134614](https://cloud.githubusercontent.com/assets/12679496/14364675/22137810-fcc7-11e5-8cda-54cfe23b3cfd.png)

**Comment by bobombolo** [comment]
just started it again with the latest master which reduces reporting and fixes the missing nodes, as well as implements the alternate block writing code supplied by @LeMagnesium. the first few lines of reporting are:

<pre>
 ./preload.py ../../worlds/mexico/map.sqlite  rasters/lq_dem_16bit_smQ.tif rasters/lq_lulc_8bit_smQ.tif 
2016-04-07 13:47:17,971 map:empty_map                :510  WARNING  WARNING: Emptying the entire map of its mapblocks
starting preload
[19.20s] [READ: 98464 pixels] [WROTE: 98504 voxels] [ETA:7365.61s]
[38.71s] [READ: 196928 pixels] [WROTE: 196996 voxels] [ETA:7405.90s]
[57.44s] [READ: 295392 pixels] [WROTE: 295525 voxels] [ETA:7306.40s]
[77.00s] [READ: 393856 pixels] [WROTE: 394014 voxels] [ETA:7327.21s]
[96.20s] [READ: 492320 pixels] [WROTE: 492492 voxels] [ETA:7303.67s]
[115.75s] [READ: 590784 pixels] [WROTE: 590971 voxels] [ETA:7304.19s]
[135.64s] [READ: 689248 pixels] [WROTE: 689443 voxels] [ETA:7317.30s]
[155.25s] [READ: 787712 pixels] [WROTE: 787915 voxels] [ETA:7308.87s]
</pre>

that's with trees turned on...

**Comment by bobombolo** [comment]
once the process finishes we should compare with realterrain mod and that other mod that moves you around the map to generate/load it: explore_map. there are issues with explore_map though where I don't think it uses the actual surface to do the teleporting but instead uses the perlin surface? anyway, worth a try for comparison.

**Comment by Tank-ACiD** [comment]
yours is actually quite faster...I'm taking 30 seconds per output line
where you're at 20, thats 33% faster..wow. what is your processor speed?

**Comment by bobombolo** [comment]
however those early estimates are way off. here is some recent output:

[8055.93s] [READ: 24911392 pixels] [WROTE: 24927916 voxels] [ETA:4191.15s]
[8106.21s] [READ: 25009856 pixels] [WROTE: 25026619 voxels] [ETA:4168.79s]
[8161.16s] [READ: 25108320 pixels] [WROTE: 25125274 voxels] [ETA:4148.59s]
[8217.39s] [READ: 25206784 pixels] [WROTE: 25223897 voxels] [ETA:4128.76s]
[8273.50s] [READ: 25305248 pixels] [WROTE: 25322549 voxels] [ETA:4108.58s]

my system is: AMD A8-5500 APU with Radeon(tm) HD Graphics × 4 and 9GB ram. SSD drive, Ubuntu latest.

**Comment by bobombolo** [comment]
my ETA calculation is wrong...

**Comment by bobombolo** [comment]
$ lscpu | grep "MHz"
CPU MHz:               2200.000
CPU max MHz:           3200.0000
CPU min MHz:           1400.0000

$ cat /proc/cpuinfo | grep "MHz"
cpu MHz		: 2600.000
cpu MHz		: 2200.000
cpu MHz		: 1800.000
cpu MHz		: 3200.000

**Comment by Tank-ACiD** [comment]
yeah im just going off elapsed/read ..the SSD is why its fast, and your
processor is faster at max than mine, and although i see all 8 of my cores
slightly active, unless the python code is specifically setup to be
multithreaded with agents and a pool, I believe it defaults to not be.

looks like you started at 20 sec per report and now around 50-55ish? I'm 5
Mil in and still holing at 30 sec.

I would imagine the jumping around to be slower as it involves graphics and
a pause after each 'landing' to let the nodes generate.

**Comment by bobombolo** [comment]
I don't know how to tell which of my processors is the one being used by python...

yes but the c++ mapgeneration seems to be lightning fast by comparisson, with even underground mapchunks of 80x80x80 taking less than one second to generate... of course it all depends on your minetest.conf render distance settings... but if we could attune the explore_map approach to move on once all generating is done, like do a teleport after each on_generated() call... we could optimize that scheme.

I am wondering why things are slowing down. I don't understand the cache in libminetest, the MapInterface.max_cache_size is a value set to 241 for some reason, and what is that? MB? does the cache just build and build until we do the save() at the end? could things go faster if we save after each row of blocks? etc...

there is of course some variability (a lot) in the ETA because I am calculating it based on elapsed/read as you are, and read is pixels while what is written varies depending on the relief in the elevation values... no easy way to predict that...

anyway.....

[10898.31s] [READ: 28948416 pixels] [WROTE: 28974598 voxels] [ETA:3359.39s]
[10994.41s] [READ: 29046880 pixels] [WROTE: 29073082 voxels] [ETA:3340.25s]
[11088.57s] [READ: 29145344 pixels] [WROTE: 29171556 voxels] [ETA:3320.01s]
[11180.19s] [READ: 29243808 pixels] [WROTE: 29270030 voxels] [ETA:3298.53s]

**Comment by Tank-ACiD** [comment]
I was about to ask why 241... Mag? That's his number right? And yes mine is
slowing down also, now 35-40 seconds btwn report line at 12.5M pix.

im sure the explore_map could be approached like a traveling salesman,
considering not only does the fov have to relocate, but also change
direction, I believe it defaults to an outward spiral, but we may want to
experiment with a moving-window and rotating fov 360 at an optimal
distance/interval.

**Comment by bobombolo** [comment]
the render distance seems to only be considered horizontally i noticed, otherwise I would suggest a drone-like down-viewing bot at an ideal distance over the surface simply doing a raster pass... not that I want to write any of that because it's such a hack...

but why is this process slowing down? I don't like that. I want to try another pass where we save to the db after each block row...

**Comment by Tank-ACiD** [comment]
correlates to the size of the sqlite dbase file, I know people complain
about slowness of files over 4GB, so maybe as its size increases so does
command latency

**Comment by bobombolo** [comment]
well, one thing, if this is the approach we use, or really, any fully generated map, is that we can start copying the sqlite file as if it is a dataset and create backups of the map that way. also for raster outputs. map.sqlite, slope.sqlite, etc.

if we multi-thread the script, one thread could be busy generating rows, another could save them to the dab, another could be... placing the nyancats... and the fourth one (i don't care about cores 5-8) could be kicking griefers. f-ing griefers... adding non-systematic error to my spatial data!

**Comment by bobombolo** [comment]
anyway, the final size of the last run of your mexico map, the map.sqlite file is only 85MB, so it's not a large file issue...

**Comment by bobombolo** [comment]
should have run cprofile on this run. maybe it's something unexpected like maybe Pillow is slower to read pixels from further down in the raster...

**Comment by bobombolo** [comment]
Done [23305.40s] sqlite file: ../../worlds/mexico/map.sqlite
total pixels read: 37871716, total surface voxels queued: 37871716, total fill voxels queued: 27371, total voxels written: 37899087, trees imported: 244900

**Comment by bobombolo** [comment]
but there is a new issue: #11

sigh

**Comment by Tank-ACiD** [comment]
mine has slowed dramatically 210sec @ 32M pixels

**Comment by Tank-ACiD** [comment]
yup yours was a lot faster prob due to processor speed

Done [40587.73s] sqlite file: map.sqlite
total pixels read: 37871716, total surface voxels queued: 37871716, total
fill v
oxels queued: 27053, total voxels written: 37898769, trees imported: 244937

**Comment by Tank-ACiD** [comment]
ok  i see the rows that have incorrect elevation, should be an easy fix,
no? still getting dark spots, not sure if I'm using mapfix or lightfix
properly though.

**Comment by bobombolo** [comment]
the fix is pushed but it was not easy. and it might be due to a lib problem so it might get fixed and break my fix. see #11 latest comments.
I am going to run your rasters through again with no trees which is much faster, and anyway the importing schemas is buggy, see #12

**Comment by bobombolo** [comment]
dark spots might never go away. they seem to appear randomly, and the faster the generating works the worse it seems to get. to use mapfix, go to the area where the darkspot is and type /mapfix

**Comment by bobombolo** [comment]
Done [7256.51s] sqlite file: ../../worlds/mexico/map.sqlite
total pixels read: 37871716, total surface voxels queued: 37871716, total fill voxels queued: 27371, total voxels written: 37899087, trees imported: 0

![screenshot_20160408_095339](https://cloud.githubusercontent.com/assets/12679496/14389430/d5668256-fd6f-11e5-8a30-94a2f8b324ce.png)

**Comment by Tank-ACiD** [comment]
I assumed it was something in Mags lib, a small workaround for now is good.
for the schemas, we probably have to do some sort of collision detection to
not overwrite existing blocks as it appears the holes are created due to
the trees having a complete bounding cubic space defined/generated? That's
just an assumption based on observation

I'm not too worried about trees right now for academic purposes, Those
can/should be generated later via point clouds.

this is leading to a much better long-term solution for real terrain
generation than the generate-on-load method I think.

**Comment by bobombolo** [comment]
the workaround is actually something I wanted to try for optimization purposes anyway, it handles the creation of the block content's internal nodeIDs more "manually" and means I have a better understanding of the minetest data format now. unfortunately it did not speed up the script much since there is multiplication involved instead of just adding to a counter but as I develop the script I think it will end up behaving much as the voxelmanipulator in lua mapgen mods does and we will iterate through all nodes in a mapblock and then a counter will work, and so we might get a performance gain thanks to this groundwork already being done. this will be important if we want below-surface-solidity.

the only downer is that having generated the whole map first it sucks to still have to explore the mape manually to simply load it into the client's view...

as for the schem import, #12 is on that, it is something that Mag should deal with since there should be options in the import_schematic function to ignore air blocks, import non-destructively, etc.

I am going to work on #5 which is to start coding symbology settings (lulc classes included) as a separate file that will make creating symbol "files" for different applications easier.

**Comment by bobombolo** [comment]
aren't trees important to the user experience of placing themselves virtually in the landscape?

**Comment by bobombolo** [comment]
as for the gis analysis, I was thinking that now that there is an in-game minimap, we could put the output of various raster analysis as optional modes in the minimap. of course this would require a discussion with core, or a hack of core. this goes towards the "separation of server/client" features discussion. I think I will move the automata remote control over to this mod if for nothing other than the features of the "teleport to surface" and set to morning buttons, and as a place to potentially create backups, import/export schemas, and possibly still to change the settings for the preload.py script to use. Point is, inside this remote, is where the output of raster analysis could be sent to an in-screen minimap, as I did in the gis mod. That was a hack of the minetest formspec, but there may now be mroe options to displaying changing image files since the introduction of the minimap. I need to find out how they accomplished this but it implies a change to what used to be an absolute impossibility of being able to display any image file that wasn't loaded from the textures folder at game launch time.

of course that does NOT allow us to visualize the output of raster analysis in 3D -- in the virtual space... so it might be an underachievement.

but in thinking about this it might make sense to simply move this python script into the realterrain mod, and the two approaches can work together. preloading could just be another option. settings and display modes could all be set with the remote, and a preload checkbox would mean that the python script is called at mod_init next time the game is loaded, which would also deal with the issue of not being able to delete or rename the map.sqlite folder while in game...

I think this bears serious consideration. I might do some inquiries and experiments tonight.

**Comment by Tank-ACiD** [comment]
i like

**Comment by bobombolo** [comment]
can you create a TIN of your research area for me to fiddle with as well? and do you have rivers and roads as a separate raster as well as in a vector format? I still haven't used any GDAL methods and would like to look at that a bit, for real world coordinates, getting information from vector sources, and testing out the use of GDAL instead of an image library in general.

**Comment by Tank-ACiD** [comment]
i can get you those 5 things tomorrow yes.

**Comment by Tank-ACiD** [comment]
let me actually make the area much smaller, i can get roads water and the
>256 elevation in a smaller tif

**Comment by Tank-ACiD** [comment]
ok, will try to email it or google drive the files,

i included a new, smaller AOI

dem is 16bit unsigned
lulcSM1 is 8bit lulc with roads and water
lulcSM2 is 8bit lulc without roads and water
lqtin_sm is a TIN based off the dem
riversSM vector
roadsSM vector

I saved the files in ArcMap so the rasters are compressed, probably want to
run them through qgis to save as uncompressed.
